src/common/ward.h

```diff
@@ -0,0 +1,63 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 Red Hat Inc.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#ifndef CEPH_COMMON_WARD_H
+#define CEPH_COMMON_WARD_H
+
+#include <utility>
+
+#include "common/backport14.h"
+
+namespace ceph {
+// Ward a scope, allow an arbitrary function to be executed on scope
+// exit. Set things up this way so we can use uncopyable/immovable functions.
+
+template<typename F>
+class warder {
```

Comment:
this duplicates `scope_guard<>` to some degree. @adamemerson can we fix `scope_guard<>` instead?